tests/test_hexary_trie.py

```diff
@@ -248,6 +254,235 @@ def test_hexary_trie_empty_squash_does_not_read_root():
+        elif len(next_change) == 2:
+            updates.append(next_change)
+            key, next_val = next_change
+            if next_val in (None, b''):
+                latest_keys.remove(key)
+        else:
+            raise Exception(f"Invalid code path: next_change = {next_change}")
+
+        # on average, build an update list of length ~100, but "shrinks" down to short lists
+        should_continue = draw(st.integers(min_value=0, max_value=99))
+        if should_continue == 0:
+            break
+
+    return inserts, updates
+
+
+@given(
+    trie_updates_strategy()
+)
+@example(
+    # Triggers a case where the delete of a leaf succeeds, but the normalization fails because
+    #   of a missing trie node. The exception was *not* preventing the pruning from happening.
+    inserts_and_updates=(
+        [
+            (b'\x00\x00\x00', b'\x00\x00\x0033333333333333333333333'),
+            (b'\x01\x00\x00', b'\x01\x00\x0033333333333333333333333'),
+        ],
+        [
+            (b'\x00\x00\x00', None),
+        ]
+    )
+)
+@example(
+    # An old implementation treated trie.set(key, b'') and trie.delete(key) differently, and
+    #   this test case exposed the issue.
+    inserts_and_updates=(
+        [],
+        [
+            (b'', b''),
+            (b'', None),
+            (b'\x00', b''),
+            (b'', b''),
+            (b'\x00', None),
+        ],
+    ),
+)
+@settings(max_examples=300)
+def test_squash_changes_does_not_prune_on_missing_trie_node(inserts_and_updates):
+    inserts, updates = inserts_and_updates
+    node_db = {}
+    trie = HexaryTrie(node_db)
+    with trie.squash_changes() as trie_batch:
+        for key, value in inserts:
+            trie_batch[key] = value
+
+    missing_nodes = dict(node_db)
+    node_db.clear()
+
+    with trie.squash_changes() as trie_batch:
+        for key, value in updates:
+            # repeat until change is complete
+            change_complete = False
+            while not change_complete:
+                # Catch any missing nodes during trie change, and fix them up.
+                # This is equivalent to Trinity's "Beam Sync".
+
+                previous_db = trie_batch.db.copy()
+                try:
+                    if value is None:
+                        del trie_batch[key]
+                    else:
+                        trie_batch[key] = value
+                except MissingTrieNode as exc:
+                    # When an exception is raised, we must never change the database
+                    current_db = trie_batch.db.copy()
+                    assert current_db == previous_db
+
+                    node_db[exc.missing_node_hash] = missing_nodes.pop(exc.missing_node_hash)
+                else:
+                    change_complete = True
+
+
+@given(
+    st.lists(st.tuples(st.binary(), st.binary())),
+    st.lists(st.binary()),
+)
+@example(
+    updates=[(b'\xa0', b'\x00\x00\x00\x00\x00\x00'), (b'\xa1', b'\x00\x00')],
```

Comment:
I'm not sure if there are useful comments that can be added to document these example cases.

Reply:
I figured out 2 of them. One of them I'm not sure. It was something that came up in the middle of implementing the pruning, not fixing an existing bug.